File: app/src/main/java/com/example/usermanagmentapp/ui/main/UserViewModel.java
package com.example.usermanagmentapp.ui.main;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.paging.Pager;
import androidx.paging.PagingData;
import androidx.paging.rxjava3.PagingRx;

import com.example.usermanagmentapp.data.model.User;
import com.example.usermanagmentapp.di.IOScheduler;
import com.example.usermanagmentapp.di.MainScheduler;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;
import io.reactivex.rxjava3.core.Scheduler;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.disposables.Disposable;
import io.reactivex.rxjava3.functions.Consumer;

/**
 * Responsible for getting user from paging data source
 */
@HiltViewModel
public class UserViewModel extends ViewModel {

    private final MutableLiveData<PagingData<User>> _state = new MutableLiveData<>();
    public final LiveData<PagingData<User>> state = _state;
    /**
     * Disposable to clear all subscriptions
     */
    private final CompositeDisposable compositeDisposable = new CompositeDisposable();

    private final Pager<Integer, User> pager;
    private final Scheduler ioScheduler;
    private final Scheduler mainScheduler;

    @Inject
    public UserViewModel(Pager<Integer, User> repository,
                         @IOScheduler Scheduler ioScheduler,
                         @MainScheduler Scheduler mainScheduler) {
        this.pager = repository;
        this.ioScheduler = ioScheduler;
        this.mainScheduler = mainScheduler;
    }

    /**
     * Subscribe to paging data source
     */
    public void pagesSubscribe() {

        Disposable disposable = PagingRx.getFlowable(pager)
                .subscribeOn(ioScheduler)
                .observeOn(mainScheduler)
                .subscribe(_state::setValue, throwable -> {

                });
        compositeDisposable.add(disposable);
    }

    /**
     * Clear all subscriptions
     */
    @Override
    protected void onCleared() {
        super.onCleared();
        compositeDisposable.clear();
    }
}

File: app/src/main/java/com/example/usermanagmentapp/ui/main/UserViewModel.kt
package com.example.usermanagmentapp.ui.main

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.paging.Pager

import androidx.paging.PagingData
import androidx.paging.PagingSource
import androidx.paging.rxjava3.flowable
import com.example.usermanagmentapp.data.model.User
import com.example.usermanagmentapp.data.repository.UserRepository
import io.reactivex.rxjava3.core.Scheduler
import io.reactivex.rxjava3.disposables.CompositeDisposable


class UserViewModel(
    private val pager: Pager<Int,User>,
    private val ioScheduler: Scheduler,
    private val mainScheduler: Scheduler,
) : ViewModel() {
    private val _pagingData = MutableLiveData<PagingData<User>>()
    val pagingData: LiveData<PagingData<User>> = _pagingData;
    private val composeDisposable = CompositeDisposable()

    fun pagesSubscribe() {

        val disposable = pager.flowable
            .subscribeOn(ioScheduler)
            .observeOn(mainScheduler)
            .subscribe({
                _pagingData.value = it
            }, {
                it.printStackTrace()
            })
        composeDisposable.add(disposable)
    }

    override fun onCleared() {
        super.onCleared()
        composeDisposable.clear()
    }
}